### PR TITLE
[C10D] Add dist.get_node_local_rank helper

### LINKED ISSUE #122816: add `torch.distributed.get_local_rank`
### 🚀 The feature, motivation and pitch

For a symmetry with `torch.distributed.get_global_rank` it would be useful to add `torch.distributed.get_local_rank` rather than have the user fish for it in the `LOCAL_RANK` env var. This feature is almost always used in dist applications.

Discussion: https://pytorch.slack.com/archives/C3PDTEV8E/p1711155684960319

In it you can see a proposal for creating a default local process group to match the `group` argument in the existing similar APIs.

Thank you.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123992

Fixes #122816

Summarizing the pros/cons of the request and motivation from #122816

- (+) it's really common for users to do 'os.getenv["LOCAL_RANK"]' so we
  should provide a helper
- (-) we can't really control if/how local rank information is made
  available, but it is handled automatically if torchrun is used.

We can assume local rank is correctly passed if it is passed.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -1120,6 +1120,28 @@ def get_pg_count() -> int:
     """
     return _world.group_count
 
+def get_node_local_rank() -> int:
+    """
+    Return the local rank of the current process relative to the node.
+
+    Semantically, this is a useful concept for mapping processes to devices.
+    For example, on a node with 8 accelerator you could use the node local rank to decide
+    which accelerator devide to bind the process to.
```
Comment (shuqiangzhang) [inline torch/distributed/distributed_c10d.py:1129]: nit: device

**Review by shuqiangzhang — APPROVED** [review]

**Review by zdevito — APPROVED** [review]
I like this because it moves in the direction of having a standard way to communicate this information rather than a mishmash of wrappers for different launchers.

**Review by XilunWu — APPROVED** [review]

**Comment by stas00** [comment]
wow, that was the fastest merge ever - haven't had a chance to give feedback.

Usually the application use pattern is:
```
int(os.environ["LOCAL_RANK"], 0)
```
so that the same code works with multi-gpu and a single gpu application.

But this implementation emits an exception, so from the framework developer point of view, one would still have to rely on `int(os.environ["LOCAL_RANK"], 0)`

Could the API be at least expanded to allow the default not to make an exception but to return 0?

**Comment by stas00** [comment]
```
def get_node_local_rank(fallback_rank=None) -> int:

    if "LOCAL_RANK" in os.environ:
        return int(os.environ["LOCAL_RANK"])
    if fallback_rank is not None:
        return fallback_rank
    raise RuntimeError(
        "LOCAL_RANK is not in the environment, and `fallback` value isn't set, so `get_node_local_rank` can't be used. "
        "Consider using torchrun or updating your process launcher to specify LOCAL_RANK."
    )


```

**Comment by wconstab** [comment]
That sounds ok to me, except- wouldn't it be hard to differentiate between the true single-gpu case (where fallback value is preferred) and the case where multi-gpu is used with an incompatible launcher?

**Comment by wconstab** [comment]
I think someone was mentioning that not all launchers set the ENV today. @zdevito @wanchaol @kurman any idea if that's true? I think @stas00's  proposal is ok actually.

**Pull request closed without merging**